modules/dmrpp_module/DmrppArray.cc

```diff
@@ -69,137 +69,827 @@ using namespace std;
 
 #define MB (1024*1024)
 #define prolog std::string("DmrppArray::").append(__func__).append("() - ")
-#define WAIT_FOR_FUTURE_MS 1
 
 namespace dmrpp {
 
-// Forward Declarations
-void *one_super_chunk_thread(void *arg_list);
-void *one_super_chunk_unconstrained_thread(void *arg_list);
 
 // ThreadPool state variables.
-std::mutex thread_pool_mtx;     // mutex for critical section
-atomic_uint thread_counter(0);
+std::mutex transfer_thread_pool_mtx;     // mutex for critical section
+atomic_uint transfer_thread_counter(0);
+
+// ThreadPool state variables.
+std::mutex compute_thread_pool_mtx;     // mutex for critical section
+atomic_uint compute_thread_counter(0);
 
 
 /**
- * @brief Uses future::wait_for() to scan the futures for a ready future. When found future::get() is called and the thead_counter is decremented.
- *
+ * @brief Uses future::wait_for() to scan the futures for a ready future, returning true when once get() has been called.
+ * When a valid, ready future is found future::get() is called and the thead_counter is decremented.
+ * Returns true when it "get"s a future (joins a thread).
+
  * @param futures The list of futures to scan
  * @param timeout The number of milliseconds to wait for each future to complete.
  * @return Returns true if future::get() was called on a ready future, false otherwise.
  */
-bool get_next_future(list<std::future<void *>> &futures, unsigned long timeout) {
-    bool joined = false;
+bool get_next_future(list<std::future<bool>> &futures, atomic_uint &thread_counter, unsigned long timeout, string parent) {
+    bool future_finished = false;
     bool done = false;
     std::chrono::milliseconds timeout_ms (timeout);
 
     while(!done){
         auto futr = futures.begin();
         auto fend = futures.end();
-        while(!joined && futr != fend){
-            // FIXME What happens if wait_for() always returns future_status::timeout for a stuck thread?
-            if((*futr).wait_for(timeout_ms) != std::future_status::timeout){
-                (*futr).get();
-                joined = true;
-                BESDEBUG(dmrpp_3, prolog << "Called future::get() on a ready future." << endl);
+        bool future_is_valid = true;
+        while(!future_finished && future_is_valid && futr != fend){
+            future_is_valid = (*futr).valid();
+            if(future_is_valid){
+                // FIXME What happens if wait_for() always returns future_status::timeout for a stuck thread?
```

Comment:
What does happen?

Reply:
I do not know

Reply:
I know what happens now: The code get's stuck in the loop endlessly waiting for the thread to finish. There is no way to directly kill the thread, but I tried using a semaphore (`atomic_bool`) and by having the code detect that the elapsed time has passed a threshold value and then sets the value of the semaphore to true. If the child thread code is written to check the value of the semaphore then it can exit by returning or throwing an exception - I tested both. Unfortunately it requires the child code to not be blocked on access to some external resource in order to be effective.